src/tools/build-manifest/src/main.rs

```diff
@@ -257,21 +257,20 @@ impl Builder {
     /// If a tool does not pass its tests, don't ship it.
     /// Right now, we do this only for Miri.
     fn check_toolstate(&mut self) {
-        let toolstates: Option<HashMap<String, String>> =
-            File::open(self.input.join("toolstates-linux.json"))
+        for file in &["toolstates-linux.json", "toolstates-windows.json"] {
+            let toolstates: Option<HashMap<String, String>> = File::open(self.input.join(file))
                 .ok()
                 .and_then(|f| serde_json::from_reader(&f).ok());
-        let toolstates = toolstates.unwrap_or_else(|| {
-            println!(
-                "WARNING: `toolstates-linux.json` missing/malformed; \
-                assuming all tools failed"
-            );
-            HashMap::default() // Use empty map if anything went wrong.
-        });
-        // Mark some tools as missing based on toolstate.
-        if toolstates.get("miri").map(|s| &*s as &str) != Some("test-pass") {
-            println!("Miri tests are not passing, removing component");
-            self.versions.disable_version(&PkgType::Miri);
+            let toolstates = toolstates.unwrap_or_else(|| {
+                println!("WARNING: `{}` missing/malformed; assuming all tools failed", file);
+                HashMap::default() // Use empty map if anything went wrong.
+            });
+            // Mark some tools as missing based on toolstate.
```

Comment:
So as implemented, this will remove Miri on all targets if any of the toolstates fails, right?

The alternative would be to only disabe Miri on Windows if Windows tests fail, and use the Linux tests for everything else.

Reply:
> So as implemented, this will remove Miri on all targets if any of the toolstates fails, right?

Correct.

> The alternative would be to only disabe Miri on Windows if Windows tests fail, and use the Linux tests for everything else.

So just leave Miri on Linux broken with `--target=*windows*`?

Reply:
> So just leave Miri on Linux broken with --target=*windows*?

That would be the consequence. OTOH, this would avoid unnecessarily breaking people that just use Miri on Linux with a Linux target.

I do not have a strong preference either way, just wanted to point out that there's a choice that we have to make here.

Reply:
> this would avoid unnecessarily breaking people that just use Miri on Linux with a Linux target (which I imagine is the majority)

That's fair. Maybe Miri can detect its Windows toolstate dynamically?

Reply:
I am not sure what that would mean, but it sounds fragile.^^

Reply:
I was thinking about packaging something like `toolstates-windows.json` along with `miri`, and let `cargo-miri` prevent using `--target=*windows*` when tests fail on Windows.

Reply:
Oh, that sounds way too complicated IMO.

Then I'd rather go with what you implemented here.^^ Just please extend the existing comments to explicitly say that we disable Miri on all targets if either of the tests failed.